x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/changepoint/SpikeAndDipDetector.java

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ml.aggs.changepoint;
+
+/**
+ * Detects spikes and dips in a time series.
+ */
+final class SpikeAndDipDetector {
+
+    private record SpikeOrDip(int index, int startExcluded, int endExcluded) {
+        double value(double[] values) {
+            return values[index];
+        }
+
+        boolean includes(int i) {
+            return i >= startExcluded && i < endExcluded;
+        }
+    }
+
+    private int argmax(double[] values, int start, int end, boolean negate) {
+        int argmax = 0;
+        double max = negate ? -values[0] : values[0];
+        for (int i = 1; i < values.length; i++) {
+            double value = negate ? -values[i] : values[i];
+            if (value > max) {
+                argmax = i;
+                max = value;
+            }
+        }
+        return argmax;
+    }
+
+    private double average(double[] values, int start, int end, boolean negate) {
+        double sum = 0.0;
+        for (int i = start; i < end; i++) {
+            sum += negate ? -values[i] : values[i];
+        }
+        return sum / (double) (end - start);
+    }
+
+    private SpikeOrDip findSpikeOrDip(double[] values, int extent, boolean negate) {
+
+        extent = Math.min(extent, values.length - 1);
+
+        final int argmax = argmax(values, 0, values.length, negate);
+
+        // Find the maximum average interval of width extent which includes argmax.
+        int maxStart = Math.max(0, argmax + 1 - extent);
+        int maxEnd = Math.min(maxStart + extent, values.length);
+        double maxAvg = average(values, maxStart, maxEnd, negate);
+        for (int start = maxStart + 1; start <= argmax; start++) {
+            if (start + extent >= values.length) {
+                break;
+            }
+            double average = average(values, start, start + extent, negate);
+            if (average > maxAvg) {
+                maxStart = start;
+                maxAvg = average;
+            }
+        }
```

Comment:
Seems weird to constantly iterate `start, ...` values when all we need is a sum & count. The count is easy as its always `extent`.

Reply:
So the idea is to find the sliding window containing the argmax on which the average is maximised (or minimised if we negate). For example, if we exclude 3 values and the sequence is ..., 1, 3, 11, 7, 5,... we want to exclude [11, 7, 5]. Alternatively, if it were ...1, 5, 11, 7, 3,... we'd want to exclude [5, 11, 7]. In fact, I could compute this in a more efficient manner, by only accounting for changes at ends, but this will be extremely fast anyway. Since maximising sum and average on a fixed length window is equivalent though I updated to avoid the division.